<a href="https://colab.research.google.com/github/yubje/AI_Practice/blob/main/tensorflow101_%EC%83%9D%ED%99%9C%EC%BD%94%EB%94%A9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. 레몬에이드 판매 예측

In [23]:
# 라이브러리 사용
import pandas as pd
import tensorflow as tf

# 데이터 읽어오기 
file_route1 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/lemonade.csv'
lemonade = pd.read_csv(file_route1)

# 데이터 모양 확인
print(lemonade.shape)
print(iris.shape)
print()

# 데이터 열 이름 확인
print(lemonade.columns)
print(iris.columns)
print()

# 데이터 상위 5개 확인 
print(lemonade.head())


(6, 2)
(150, 5)

Index(['온도', '판매량'], dtype='object')
Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종'], dtype='object')

   온도  판매량
0  20   40
1  21   42
2  22   44
3  23   46
4  24   48

       nox  lstat  medv
0    0.538   4.98  24.0
1    0.469   9.14  21.6
2    0.469   4.03  34.7
3    0.458   2.94  33.4
4    0.458   5.33  36.2
..     ...    ...   ...
501  0.573   9.67  22.4
502  0.573   9.08  20.6
503  0.573   5.64  23.9
504  0.573   6.48  22.0
505  0.573   7.88  11.9

[506 rows x 3 columns]


In [19]:
# 독립 변수와 종속 변수 분리 
independent = lemonade[['온도']]
dependent = lemonade[['판매량']]

print(independent)
print(dependent)

   온도
0  20
1  21
2  22
3  23
4  24
5  25
   판매량
0   40
1   42
2   44
3   46
4   48
5   50


In [20]:
# 모델의 구조를 만든다. 

# 독립변수의 개수 1개
X = tf.keras.layers.Input(shape=[1])
# 종속변수의 개수 1개 
Y = tf.keras.layers.Dense(1)(X)

# 독립변수, 종속변수
model = tf.keras.models.Model(X, Y)
# 컴파일 
model.compile(loss='mse')

## 모델 학습

### loss
- (예측 - 결과)^2

- 그 시점의 모델이 얼마나 정답에 가깝게 맞추고 있는지 평가하는 지표 

- 종속 변수와 예측 결과의 차이를 계산해서 그 값들의 평균을 구한 것

- loss가 0에 가까워질수록 정답에 가까움


### 진행 횟수
- Epoch 2/100 : 100회의 학습 중 2번째 학습

### 진행 속도
- 5ms/step : 한 회당 학습 속도

In [27]:
# 데이터로 모델을 학습(FIT)한다.
# epochs: 몇 번 학습할 것인지 입력
# verbose=0 : 화면 출력 X
model.fit(independent, dependent, epochs=990, verbose=0)
model.fit(independent, dependent, epochs=10)

Epoch 1/10
1/1 [==============================] - 0s 4ms/step - loss: 3.7990e-04
Epoch 2/10
1/1 [==============================] - 0s 8ms/step - loss: 3.7966e-04
Epoch 3/10
1/1 [==============================] - 0s 5ms/step - loss: 3.7948e-04
Epoch 4/10
1/1 [==============================] - 0s 6ms/step - loss: 3.7923e-04
Epoch 5/10
1/1 [==============================] - 0s 7ms/step - loss: 3.7909e-04
Epoch 6/10
1/1 [==============================] - 0s 5ms/step - loss: 3.7897e-04
Epoch 7/10
1/1 [==============================] - 0s 8ms/step - loss: 3.7879e-04
Epoch 8/10
1/1 [==============================] - 0s 8ms/step - loss: 3.7855e-04
Epoch 9/10
1/1 [==============================] - 0s 3ms/step - loss: 3.7833e-04
Epoch 10/10
1/1 [==============================] - 0s 7ms/step - loss: 3.7807e-04


In [28]:
# 모델을 이용한다
model.predict(independent)


array([[40.034195],
       [42.02566 ],
       [44.01713 ],
       [46.008602],
       [48.00007 ],
       [49.99154 ]], dtype=float32)

In [10]:
print("Predictions: ", model.predict([[15]]))

Predictions:  [[22.990213]]


# 2. 보스턴 집값 예측

## 구조
- 13개의 독립변수와 1개의 독립변수
- y = w1*x1 + w2*x2 + ... + w13*x13 + b  (퍼셉트론)
- Weight(가중치) : w1, w2, ,,,
- Bias(편향) : b



## 각 열 의미
- medv: 중간값(집값), 단위 $1000
- crim: 범죄율
- zn
- indus
- chas: 강변(charles river)
- nox
- rm: 평균 방 수
- age: 노후주택 비율
- dis 
- rad
- tax: 재산세 비율
- ptratio: 학생/교사 비율
- b
- lstat: 하위계층 비율 



In [2]:
# 라이브러리 사용
import tensorflow as tf
import pandas as pd

# 1. 데이터 준비
file_route2 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
boston = pd.read_csv(file_route2)

# 1-1. 준비한 데이터 확인 
print("데이터 확인")
print(boston.shape)
print(boston.columns)
print(boston.head())

# 칼럼 선택하기
print(boston[['nox', 'lstat', 'medv']])

# 1-2. 독립변수, 종속변수 분리
independent = boston[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
            'ptratio', 'b', 'lstat']]
dependent = boston[['medv']]
print(independent.head())
print(dependent.head())

ModuleNotFoundError: No module named 'tensorflow'

In [11]:
# 2. 모델의 구조를 만든다.
X = tf.keras.layers.Input(shape=[13])
Y = tf.keras.layers.Dense(1)(X)
model = tf.keras.models.Model(X, Y)
model.compile(loss='mse')

## 독립변수가 12개, 종속변수가 2개라면?

### 코드
```
X = tf.keras.layers.Input(shape=[12])
Y = tf.keras.layers.Dense(2)(X)
model = tf.keras.models.Model(X, Y)
model.compile(loss='mse')
```

### 수식
```
y1 = w1*x1 + w2*x2 + ... + w12*x12 + b1
y2 = w'1*x1 + w'2*x2 + ... + w'12*x12 + b2
```

In [14]:
# 3. 데이터로 모델을 학습(FIT)한다.
model.fit(independent, dependent, epochs=1000, verbose=0)
model.fit(independent, dependent, epochs=10)

Epoch 1/10
16/16 [==============================] - 0s 609us/step - loss: 23.8169
Epoch 2/10
16/16 [==============================] - 0s 773us/step - loss: 23.7324
Epoch 3/10
16/16 [==============================] - 0s 686us/step - loss: 23.8459
Epoch 4/10
16/16 [==============================] - 0s 740us/step - loss: 24.1948
Epoch 5/10
16/16 [==============================] - 0s 800us/step - loss: 23.7135
Epoch 6/10
16/16 [==============================] - 0s 693us/step - loss: 23.7262
Epoch 7/10
16/16 [==============================] - 0s 725us/step - loss: 23.9809
Epoch 8/10
16/16 [==============================] - 0s 790us/step - loss: 24.0101
Epoch 9/10
16/16 [==============================] - 0s 810us/step - loss: 23.9139
Epoch 10/10
16/16 [==============================] - 0s 825us/step - loss: 24.1359


In [15]:
# 4. 모델을 이용한다. 
print(model.predict(independent[5:10]))
print(dependent[5:10])

[[24.767962]
 [21.233839]
 [18.217527]
 [10.068386]
 [17.597456]]
   medv
5  28.7
6  22.9
7  27.1
8  16.5
9  18.9


In [16]:
# 모델 수식 확인
print(model.get_weights())

[array([[-0.09379134],
       [ 0.04781507],
       [-0.02313689],
       [ 2.651204  ],
       [-1.8426425 ],
       [ 5.2640595 ],
       [-0.0117174 ],
       [-1.060684  ],
       [ 0.1982418 ],
       [-0.01118985],
       [-0.5172329 ],
       [ 0.0118592 ],
       [-0.46270436]], dtype=float32), array([8.23354], dtype=float32)]


house_price = -0.09296059 * x1 + 0.04728362 * x2 + -0.0201432 * x3 + ...


# 3. 아이리스 품종 분류
- 범주형 데이터
- 분류(Classification)

### 범주형 데이터 vs 양적 데이터
- 분류(classification): 종속변수가 범주형 데이터 ex. 아이리스 품종 분류
- 회귀(regression) : 종속변수가 양적 데이터 ex. 레모네이드 판매 예측, 보스턴 집값 예측

In [1]:
# 1. 데이터 가져오기
# 라이브러리 사용
import tensorflow as tf
import pandas as pd


# 데이터 읽어오기 

file_route3 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
iris = pd.read_csv(file_route3)

# 데이터 모양 확인
print("데이터 모양")
print(iris.shape)
print()

# 데이터 열 이름 확인
print("데이터 열 이름")
print(iris.columns)
print()

# 데이터 상위 5개 확인
print("데이터 상위 5개")
print(iris.head())
print()

# onehot-encoding: 품종(setosa, virginica, versicolor) => '품종_setosa', '품종_versicolor', '품종_virginica'
# 3가지 품종을 3개 열로 분리 => 종속변수 3개
iris = pd.get_dummies(iris)


independent = iris[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
dependent = iris[['품종_setosa', '품종_versicolor', '품종_virginica']]

print("독립변수")
print(independent.head())
print()


print("종속변수")
print(dependent.head())
print()

데이터 모양
(150, 5)

데이터 열 이름
Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종'], dtype='object')

데이터 상위 5개
   꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

독립변수
   꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭
0   5.1  3.5    1.4   0.2
1   4.9  3.0    1.4   0.2
2   4.7  3.2    1.3   0.2
3   4.6  3.1    1.5   0.2
4   5.0  3.6    1.4   0.2

종속변수
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0



## softmax & crossentropy

### softmax 
- 회귀분석의 경우 종속변수의 값의 크기에 제한이 없지만, 분류의 경우 종속변수의 값이 0 ~ 1에 한정되어 있다.
- softmax는 종속변수의 값을 0 ~ 1 범주로 가공시키는 함수 

### crossentropy
- loss값이 crossentropy를 사용해 계산되도록 함 

In [ ]:
# 2. 모델의 구조를 만든다.
X = tf.keras.layers.Input(shape=[4])

# Activation함수(활성화 함수)
# sigmoid, softmax: 특정 결과가 나올 확률을 수치로 도출 
# softmax: 0과 1 사이의 결과값으로 나옴 
Y = tf.keras.layers.Dense(3, activation='softmax')(X)
model = tf.keras.models.Model(X, Y)
# categorical_crossentropy로 계산된 차이값
# accuracy를 같이 보여줌 
model.compile(loss='categorical_crossentropy', metrics='accuracy')

In [ ]:
# 3. 데이터로 모델을 학습(Fit)합니다.
model.fit(independent, dependent, epochs=1000, verbose=0)
model.fit(independent, dependent, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.1640 - accuracy: 0.9733
Epoch 2/10
5/5 [==============================] - 0s 1ms/step - loss: 0.1639 - accuracy: 0.9733
Epoch 3/10
5/5 [==============================] - 0s 1ms/step - loss: 0.1634 - accuracy: 0.9733
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.1638 - accuracy: 0.9800
Epoch 5/10
5/5 [==============================] - 0s 1ms/step - loss: 0.1630 - accuracy: 0.9733
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.1631 - accuracy: 0.9733
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.1639 - accuracy: 0.9733
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.1632 - accuracy: 0.9800
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.1626 - accuracy: 0.9733
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.1625 - accuracy: 0.9800


In [ ]:
# 4. 모델을 이용한다.
# 첫 데이터 5개
print(model.predict(independent[:5]))
print(dependent[:5])

print()
# 마지막 데이터 5개
print(model.predict(independent[-5:]))
print(dependent[-5:])

[[9.9917406e-01 8.2578720e-04 6.1443565e-08]
 [9.9729484e-01 2.7046339e-03 5.2496580e-07]
 [9.9837565e-01 1.6240928e-03 2.1870348e-07]
 [9.9560285e-01 4.3961853e-03 9.9022327e-07]
 [9.9923921e-01 7.6082471e-04 5.0672131e-08]]
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0

[[4.8320834e-07 1.3877459e-01 8.6122489e-01]
 [1.2641316e-06 1.8540102e-01 8.1459767e-01]
 [1.9838249e-06 2.5119054e-01 7.4880743e-01]
 [3.2583685e-07 1.5083568e-01 8.4916395e-01]
 [3.7356831e-06 3.2534212e-01 6.7465419e-01]]
     품종_setosa  품종_versicolor  품종_virginica
145          0              0             1
146          0              0             1
147          0              0             1
148          0              0             1
149          0              0             1


In [ ]:
# weights & bias 출력
print(model.get_weights())

[array([[ 1.3729701 ,  0.2611539 , -0.31711322],
       [ 2.2308123 ,  0.29900244, -1.3873504 ],
       [-3.9313948 , -0.4603671 ,  0.979452  ],
       [-4.329924  , -0.85880154,  1.9708254 ]], dtype=float32), array([ 1.6765242,  1.4561107, -1.7801377], dtype=float32)]


# 4. 히든레이어

## 1) 보스턴 집값 예측 
- Input Layer : x1, x2, x3, ... , x13
- Hidden Layer : h1, h2, h3, h4, h5
- Output Layer : y
- (506, 13) -> (506, 5) -> (506, 1)

In [ ]:
# 라이브러리 사용
import tensorflow as tf
import pandas as pd
 

# 1.데이터 준비
file_route2 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
boston = pd.read_csv(file_route2)

# 종속변수, 독립변수
independent = boston[['crim', 'zn', 'indus', 'chas', 'nox', 
            'rm', 'age', 'dis', 'rad', 'tax',
            'ptratio', 'b', 'lstat']]
dependent = boston[['medv']]
print(independent.shape, dependent.shape)

In [1]:
# 2-2. 모델 구조 - hidden layer 1개
X = tf.keras.layers.Input(shape=[13])
H = tf.keras.layers.Dense(10, activation='swish')(X)  # activation : 활성화 함수 
Y = tf.keras.layers.Dense(1)(H)
model = tf.keras.models.Model(X, Y)
model.compile(loss='mse')

# 모델 구조 확인
model.summary()

NameError: name 'tf' is not defined

In [ ]:
# 2-2. 모델 구조 - hidden layer 3개
X = tf.keras.layers.Input(shape=[13])
H = tf.keras.layers.Dense(5, activation='swish')(X)  # activation : 활성화 함수 
H = tf.keras.layers.Dense(3, activation='swish')(X) 
H = tf.keras.layers.Dense(3, activation='swish')(X) 
Y = tf.keras.layers.Dense(1)(H)
model = tf.keras.models.Model(X, Y)
model.compile(loss='mse')

# 모델 구조 확인
model.summary()

In [ ]:
# 3.데이터로 모델 학습(FIT)
model.fit(independent, dependent, epochs=100)

In [ ]:
# 4. 모델 이용
print(model.predict(independent[:5]))
print(dependent[:5])
 

# 모델의 수식 확인
print(model.get_weights())

## 2) 아이리스 품종 분류

In [ ]:
# 1. 데이터 가져오기
# 라이브러리 사용
import tensorflow as tf
import pandas as pd


# 데이터 읽어오기 
file_route3 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
iris = pd.read_csv(file_route3)

# 데이터 모양 확인
print("데이터 모양")
print(iris.shape)
print()

# 데이터 열 이름 확인
print("데이터 열 이름")
print(iris.columns)
print()

# 데이터 상위 5개 확인
print("데이터 상위 5개")
print(iris.head())
print()

# onehot-encoding: 품종(setosa, virginica, versicolor) => '품종_setosa', '품종_versicolor', '품종_virginica'
# 3가지 품종을 3개 열로 분리 => 종속변수 3개
iris = pd.get_dummies(iris)

independent = iris[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
dependent = iris[['품종_setosa', '품종_versicolor', '품종_virginica']]

print("독립변수")
print(independent.head())
print(independent.shape)
print()

print("종속변수")
print(dependent.head())
print(dependent.shape)
print()